### ISSUE TITLE
Make validate_params work with the function's parametters instead of a dict

### ISSUE THREAD

**Issue by Seluj78** [Seluj78/flask-utils#30]
Instead of having

```py
@validate_params(
    {
        "name": str,
        "age": int,
        "is_student": bool,
        "courses": List[str],
        "grades": Dict[str, int],
    }
)
def example():
```

It would make more sense to have the decorator read the function arguments directly like so :

```py
@validate_params
def example(name: str, age: int, is_student: bool, courses: List[str], grades: Dict[str, int]):
```

/!\ One attention point is to see how Flask interacts with the function and where it places it's variables when doing a function with some kind of slug (`@app.route("/users/<user_id>")`)

**Comment by Seluj78** [comment]
Credit for this idea goes to @JulienPalard

**Comment by Seluj78** [comment]
> I'd add: choose another name for this decorator. Both for backward compatibility and to try to stuff more meaning in it?

@JulienPalard that's a good idea, do you have any recommendations ? :) I was thinking `validate_arguments` but I'm not sure

**Comment by Mews** [comment]
I think I might have an idea on how to accomplish this ngl.
Are there tests in place to test the decorator so I can make sure I dont break anything?

**Comment by Seluj78** [comment]
Yes there are ! See the tests folder, they run with `pytest` and `tox`

Check the readme (I think I detailed it here) or the contributing file and if it's not detailed let me know and I'll add it

**Comment by Mews** [comment]
Alright thanks 👍 
I know how to run the tests haha I was just wondering if this particular feature was accounted for in them.
Btw, another question, since I'm not too familiar with flask.
Take, for example the example you provided:
```py
@validate_params
def example(name: str, age: int, is_student: bool, courses: List[str], grades: Dict[str, int]):
    ...
```
Would the arguments `name`, `age` etc be used inside the function? Or would they be fetched from the json?
(Again, this is based only on the context I got from looking at the source code so I might be misinterpreting what it's doing)

**Comment by Mews** [comment]
Also I seem to be running into some issues using tox to run the tests, but I can run them fine with pytest.
Is it fine to run the tests with only pytest or should I look into fixing tox.

**Comment by Seluj78** [comment]
As for your question about the function, it's a little more complicated than that.

Flask can also inject some variables in the function when using slugs (for example https://stackoverflow.com/a/40460690)

And then we want to validate the JSON (and maybe one day form data) params that would be inputted along the potentials slugs

(Sorry I'm on my phone I can't give you a better example right now)

**Comment by Seluj78** [comment]
@all-contributors please add @JulienPalard for ideas

**Comment by Seluj78** [comment]
@Mews let me know if you have more questions for this :)

**Comment by Mews** [comment]
Oh yeah sorry I still do plan on contributing to this issue I was just busy with something else.
So is there a reason to pass `*args, **kwargs` to the decorated function here?
https://github.com/Seluj78/flask-utils/blob/main/flask_utils/decorators.py#L274

**Comment by Seluj78** [comment]
> Oh yeah sorry I still do plan on contributing to this issue I was just busy with something else.

No worries, take your time :)
> So is there a reason to pass `*args, **kwargs` to the decorated function here? [`main`/flask_utils/decorators.py#L274](https://github.com/Seluj78/flask-utils/blob/main/flask_utils/decorators.py?rgh-link-date=2024-06-13T08%3A30%3A24Z#L274)


Yes there is :) Because of the way flask work, there already are args and kwargs to the function so we need to pass them along :)

**Comment by Mews** [comment]
Okay and the name, age etc arguments should also be passed to the function yes?
If so I can probably get to working on it right away.

**Comment by Mews** [comment]
Okay, but I'll probably need to change the tests too right? To use the new syntax you proposed

**Comment by Seluj78** [comment]
Yes that's correct ! But don't worry too much about tests and documentation for now, at least not before we validated it does what we want

**Comment by Mews** [comment]
Also, taking a look at the tests I found this:
https://github.com/Seluj78/flask-utils/blob/main/tests/test_validate_params.py#L247
However, for function arguments you can't use type hints in the same way
```
# Not a valid type hint
def example(name: (str, int)):
    etc...
```
Python does evaluate it but it's not a very good way to do it.
The way to indicate it would be through `typing.Tuple[str, int]` or through `typing.Union[str, int]`.
My question is just what would you prefer to be used? I'm leaning more to `typing.Tuple` because its easier to tell what it means but idk

**Comment by Mews** [comment]
Also I have a question about a test that's being ran, here:
https://github.com/Seluj78/flask-utils/blob/3ec12cf62f6112cbd16ef2a656993f7e337ee976/tests/test_validate_params.py#L243-L256
Maybe I'm wrong, but in `test_valid_request` aren't you passing a dict, which would be an invalid type?

**Comment by Mews** [comment]
Sorry for spamming this thread but I've began implementing a few of the changes necessary and I have a question related to pull requests and stuff. I'm making many changes on the same file, on many parts of the code. My question is, should I make a commit for each change individually, or make a single big commit when the entire file is edited?

**Comment by JulienPalard** [comment]
> My question is, should I make a commit for each change individually, or make a single big commit when I'm done editing the file?

My point of view for this is: each commit should work Or not break the whole project. Read: tests shall pass.

Idea is that I see `git commits` like saving in a video game: do you save a single time at the end of a hard dungeon, or do you save each time you complete one room? (Or do you save when you're in a half-backed/bad-position? NO? Same with code, don't commit it).

Another more distant argument is: is someone want to use `git bisect`, he does not want to hit commit known to be fully broken.

I also abuse of `git rebase -i` in case I have "too small" commits like "ok I need to sleep, this is half backed, but still I'm like: `git commit -m wip; git push` in case of fire, so I often have to "clean the mess up for the history", you don't need to play like this ;)

An approach that is slowly maturing in me is to write the commit message **before** doing the work, it's a clean way to delimit the task to be done AND to write good commit messages, let's call it "commit message driven development".

**Comment by Mews** [comment]
Hi @Seluj78
Apart from that one `TupleUnion` test, this feature is practically done. 
There's only one issue left, which is about optional arguments.
Currently, I'm passing the arguments from the json to the decorated function like this:
```py
for key in data:
    kwargs[key] = data[key]

return fn(*args, **kwargs)
```
However, this causes an obvious issue, which is that if a parameter is optional and is not present in `data`, no error will be raised, since its optional, but it also wont be added to `kwargs` and as such not passed to the wrapped function, which would raise `TypeError: missing x required positional arguments`.

I was thinking you could get around this by doing:
```py
for key in parameters:
    if _is_optional(parameters[key]):
        kwargs[key] = None

    else:
        kwargs[key] = data[key]
```
So, when an optional parameter isn't passed in the json, it gets passed to the function as `None`. This passes all the tests (after changing them to the new syntax).
What do you think of this?

**Comment by JulienPalard** [comment]
What about making optional arguments optional in the prototype too?

àla :

```python
@validate_params
def example(name: str, age : Optional[int] = None):
    ...
```

**Comment by JulienPalard** [comment]
```python
for key in data: kwargs[key] = data[key]
```

I don't have the context, but please take extra care (by adding a test or two) of not having a query string argument clobbered by a value from the request body.

I mean (I'm no flask user forgive inconsistencies):

```
@validate_params
@app.route("/user/{user_id}")
def get_user(user_id, is_admin: bool):
    ...
```

If someone sends:

    curl -XPOST -H "Content-Type: application/json" -d '{"user_id": 666, "is_admin": False}' localhost:8000/user/42

I except to get `user_id` equal to `42`, not `666`.

**Comment by Mews** [comment]
> ```python
> for key in data: kwargs[key] = data[key]
> ```
> 
> I don't have the context, but please take extra care (by adding a test or two) of not having a query string argument clobbered by a value from the request body.
> 
> I mean (I'm no flask user forgive inconsistencies):
> 
> ```
> @validate_params
> @app.route("/user/{user_id}")
> def get_user(user_id, is_admin: bool):
>     ...
> ```
> 
> If someone sends:
> 
> ```
> curl -XPOST -H "Content-Type: application/json" -d '{"user_id": 666, "is_admin": False}' localhost:8000/user/42
> ```
> 
> I except to get `user_id` equal to `42`, not `666`.

I mean I'm not a flask expert either, that's why I asked @Seluj78 if the way I was doing it wasn't gonna cause issues. I guess we'll have to wait until he responds.

**Comment by Seluj78** [comment]
(I'm not dead, just was really busy these past few days. I will respond ASAP, probably tomorrow)

**Comment by Mews** [comment]
No worries haha

**Comment by Seluj78** [comment]
For the tuple and union question, I do agree with you.

We shouldn't allow the usage of

```py
# Not a valid type hint
def example(name: (str, int)):
    etc...
```

What we can probably do, I don't know how though, is raise a DeprecationWarning when we see this usage.

As for the choice between Tuple and Union, I agree with you. `typing.Tuple` should be used because it means that a tuple is passed (for example `Tuple[str, int]` -> (`42, 'hello')`) whereas `typing.Union` means that one of the types is passed (for example `Union[str, int]` -> `42` or `'hello'`).

For your question about the `test_valid_request` for the `TestTupleUnion` class, I agree with you. It should raise an error if we implement the above thing where it's either a tuple passed to the function (works in Python functions, not with JSON data)

So the test should test for `typing.Tuple`, `typing.Union` and the normal Python tuple `(Any, Any, ...)`

As for the commit messages question, I agree with Julien. I also do try to keep my commit messages clean and each commit working when I can !

Do let me know if you have more questions. In any case I will test your PR and code once you have it pushed so we can check stuff together.

**Comment by Mews** [comment]
I was looking into that test and I think I interpreted what its for incorrectly lol.
Its not for validating that the tuple contains a string and an int, it indicates an Union type where the argument can be either a string or a int. Still, when defining type hints it would be interpreted differently, so in my opinion it should just be removed.

My proposal for allowing validating tuples still stands though and I'd be happy to introduce it too :)

**Comment by Seluj78** [comment]
> My proposal for allowing validating tuples still stands though and I'd be happy to introduce it too :)

Works for me !
